File: src/main/java/org/lanternpowered/server/util/palette/PaletteBasedArray.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.util.palette;

import org.lanternpowered.server.util.collect.array.VariableValueArray;

public interface PaletteBasedArray<T> {

    /**
     * Gets the capacity.
     *
     * @return The capacity
     */
    int getCapacity();

    /**
     * Gets the {@link T} at the given index.
     *
     * @param index The index
     * @return The object
     */
    T get(int index);

    /**
     * Sets the {@link T} at the given index.
     *
     * @param index The index
     * @param object The object
     * @return The previous assigned value
     */
    T set(int index, T object);

    /**
     * Gets the {@link Palette} of this palette based object
     * array. Does not allow removal of {@link T}s.
     *
     * @return The object palette
     */
    Palette<T> getPalette();

    /**
     * Gets the backing {@link VariableValueArray} which
     * holds the integer ids.
     *
     * @return The backing ids array
     */
    VariableValueArray getBacking();

    /**
     * Creates a copy of this {@link PaletteBasedArray}.
     *
     * @return The copy
     */
    PaletteBasedArray<T> copy();
}

File: src/main/kotlin/org/lanternpowered/api/util/palette/PaletteBasedArray.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.api.util.palette

import org.lanternpowered.api.util.VariableValueArray

/**
 * Represents an object array that stores objects
 * using a [VariableValueArray] and a [Palette].
 */
interface PaletteBasedArray<T : Any> {

    /**
     * The size.
     */
    val size: Int

    /**
     * Gets the [T] at the given index.
     *
     * @param index The index
     * @return The object
     */
    operator fun get(index: Int): T

    /**
     * Sets the [T] at the given index.
     *
     * @param index The index
     * @param obj The object
     * @return The previous assigned value
     */
    operator fun set(index: Int, obj: T)

    /**
     * Gets the [Palette] of this palette based object
     * array. Does not allow removal of [T]s.
     *
     * @return The object palette
     */
    val palette: Palette<T>

    /**
     * The backing [VariableValueArray] which holds the integer ids.
     */
    val backing: VariableValueArray

    /**
     * Creates a copy of this [PaletteBasedArray].
     *
     * @return The copy
     */
    fun copy(): PaletteBasedArray<T>

    /**
     * Serializes this [PaletteBasedArray].
     */
    fun serialize(): SerializedPaletteBasedArray<T>
}
